File: Frontend/app/api/speech-to-text-async/route.ts
import { NextRequest, NextResponse } from "next/server";
import OpenAI from "openai";
import { toFile } from "openai/uploads";

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

export async function POST(request: NextRequest) {
  try {
    if (!process.env.OPENAI_API_KEY) {
      return NextResponse.json(
        { error: "OpenAI API key not configured" },
        { status: 500 }
      );
    }

    const formData = await request.formData();

    // Get audio file — look for "audio" key first, then any File in the form
    let audioFile = formData.get("audio") as File | null;

    if (!audioFile || !(audioFile instanceof File)) {
      for (const [, val] of formData.entries()) {
        if (val instanceof File) {
          audioFile = val;
          break;
        }
      }
    }

    if (!audioFile) {
      return NextResponse.json(
        { error: "No audio file provided" },
        { status: 400 }
      );
    }

    console.log(
      `[STT] Received: name="${audioFile.name}", size=${(audioFile.size / 1024).toFixed(2)} KB, type="${audioFile.type}"`
    );

    if (audioFile.size < 1000) {
      return NextResponse.json(
        { error: "Audio file too small — likely no audio captured" },
        { status: 400 }
      );
    }

    // Read raw bytes and wrap as a proper File for the OpenAI SDK
    const arrayBuffer = await audioFile.arrayBuffer();
    const file = await toFile(Buffer.from(arrayBuffer), "audio.webm", {
      type: "audio/webm",
    });

    console.log(`[STT] Sending ${(arrayBuffer.byteLength / 1024).toFixed(2)} KB to gpt-4o-mini-transcribe...`);

    // Use OpenAI SDK — sends audio directly to transcription, no conversion
    const transcription = await openai.audio.transcriptions.create({
      file,
      model: "gpt-4o-mini-transcribe",
    });

    console.log("[STT] Transcription result:", transcription.text);

    if (!transcription.text || !transcription.text.trim()) {
      return NextResponse.json(
        { error: "No speech detected" },
        { status: 400 }
      );
    }

    return NextResponse.json({
      text: transcription.text.trim(),
      processingMethod: "gpt-4o-mini-transcribe",
    });

  } catch (err: any) {
    console.error("[STT] Error:", err);
    return NextResponse.json(
      { error: err.message || "Transcription failed" },
      { status: 500 }
    );
  }
}

File: Backend/stt/route.py
import os
import io
from fastapi import APIRouter, Depends, File, HTTPException, UploadFile
from fastapi.responses import JSONResponse

from openai import OpenAI

from utils.auth import (
    RequestAuth,
    get_request_auth_required,
)
from utils.redis_limiter import check_rate_limit

router = APIRouter()

client = OpenAI(
    api_key=os.getenv("OPENAI_API_KEY")
)

MAX_AUDIO_SIZE = 25 * 1024 * 1024  # 25 MB


@router.post("/speech-to-text")
async def speech_to_text(
    audio: UploadFile = File(...),
    auth_ctx: RequestAuth = Depends(get_request_auth_required),
):
    
   
    await check_rate_limit(
        user_id=auth_ctx.user_id,
        endpoint="speech-to-text"
    )

    
    if not os.getenv("OPENAI_API_KEY"):
        raise HTTPException(
            status_code=500,
            detail="OpenAI API key not configured"
        )


    if not audio:
        raise HTTPException(
            status_code=400,
            detail="No audio file provided"
        )

    file_bytes = await audio.read()

    file_size = len(file_bytes)

    print(
        f"[STT] Received: "
        f'name="{audio.filename}", '
        f"size={file_size / 1024:.2f} KB, "
        f'type="{audio.content_type}"'
    )

    if file_size < 1000:
        raise HTTPException(
            status_code=400,
            detail="Audio file too small — likely no audio captured"
        )

    if file_size > MAX_AUDIO_SIZE:
        raise HTTPException(
            status_code=413,
            detail="Audio exceeds 25MB limit"
        )

    
    await audio.seek(0)

    print(
        f"[STT] Sending {file_size / 1024:.2f} KB to GPT-4o Mini Transcribe..."
    )

   
    buffer = io.BytesIO(file_bytes)
    buffer.name = audio.filename or "audio.webm"
    transcription = client.audio.transcriptions.create(
        model="gpt-4o-mini-transcribe",
        file=audio.file,
    )

    print("[STT] Transcription:", transcription.text)

    if not transcription.text or not transcription.text.strip():
        raise HTTPException(
            status_code=400,
            detail="No speech detected"
        )

    return JSONResponse(
        {
            "text": transcription.text.strip(),
            "processingMethod": "gpt-4o-mini-transcribe",
        }
    )